src/oid4vp/holder.rs

```diff
@@ -0,0 +1,397 @@
+        let request = self
+            .validate_request(url)
+            .await
+            .map_err(|e| OID4VPError::RequestValidation(format!("{e:?}")))?;
+
+        match request.response_mode() {
+            ResponseMode::DirectPost | ResponseMode::DirectPostJwt => {
+                self.permission_request(request).await
+            }
+            ResponseMode::Unsupported(mode) => {
+                Err(OID4VPError::UnsupportedResponseMode(mode.to_owned()))
+            }
+        }
+    }
+
+    pub async fn submit_permission_response(
+        &self,
+        response: Arc<PermissionResponse>,
+    ) -> Result<Option<Url>, OID4VPError> {
+        // Create a descriptor map for the presentation submission based on the credentials
+        // returned from the selection response.
+        let Some(input_descriptor_id) = response
+            .presentation_definition
+            .input_descriptors()
+            .first()
+            .map(|d| d.id().to_owned())
+        else {
+            // NOTE: We may wish to add a more generic `BadRequest` error type
+            // we should always expect to have at least one input descriptor.
+            return Err(OID4VPError::InputDescriptorNotFound);
+        };
+
+        let descriptor_map =
+            self.create_descriptor_map(&response.selected_credential, input_descriptor_id);
+
+        let presentation_submission_id = uuid::Uuid::new_v4();
+        let presentation_definition_id = response.presentation_definition.id().clone();
+
+        // // Create a presentation submission.
+        let presentation_submission = PresentationSubmission::new(
+            presentation_submission_id,
+            presentation_definition_id,
+            vec![descriptor_map],
+        );
+
+        let vp_token = self
+            .create_verifiable_presentation(&response.selected_credential)
+            .await?;
+
+        let response = self
+            .submit_response(
+                response.authorization_request.clone(),
+                AuthorizationResponse::Unencoded(UnencodedAuthorizationResponse(
+                    Default::default(),
+                    vp_token,
+                    presentation_submission,
+                )),
+            )
+            .await
+            .map_err(|e| OID4VPError::ResponseSubmission(format!("{e:?}")))?;
+
+        Ok(response)
+    }
+}
+
+// Internal methods for the Holder.
+impl Holder {
+    /// Return the static metadata for the holder.
+    ///
+    /// This method is used to initialize the metadata for the holder.
+    pub(crate) fn metadata() -> Result<WalletMetadata, OID4VPError> {
+        let mut metadata = WalletMetadata::openid4vp_scheme_static();
+
+        // Insert support for the VCDM2 SD JWT format.
+        metadata.vp_formats_supported_mut().0.insert(
+            ClaimFormatDesignation::Other("vcdm2_sd_jwt".into()),
+            ClaimFormatPayload::AlgValuesSupported(vec!["ES256".into()]),
+        );
```

Comment:
I think our library should be extensible so that we don't need to alter the Rust code for every new vp format that we need to support in the native libraries.

Reply:
Okay to track this in another PR for metadata manager? I'll try to revive the old code from a month or so ago and make this more extensible.

Reply:
tracking in https://github.com/spruceid/mobile-sdk-rs/pull/64

Reply:
Fine to merge without this change.